pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -272,10 +273,32 @@ func (esc *endpointSliceController) updateEndpointCacheForSlice(hostName host.Na
 				}
 
 				istioEndpoint := builder.buildIstioEndpoint(a, portNum, portName, discoverabilityPolicy, healthStatus)
-				endpoints = append(endpoints, istioEndpoint)
+				if features.EnableDualStack && (pod != nil) && len(pod.Status.PodIPs) > 1 {
+					var addrs []string
+					for _, addr := range pod.Status.PodIPs {
+						addrs = append(addrs, addr.IP)
+					}
```

Comment:
Confused why we loop over Address *and* PodIPs.

For a dual stack pod, Address will already container 2 entries (v4 and v6). Now we make `addrs` the PodIPs (also v4 and v6), but then append them.

Won't this result in 4 addresses instead of 2?

Again, there seems to be no tests at all for this.

Reply:
@zhlsunshine : could you please confirm that for dual-stack, we get two separate slice.EndPoints, each of them with a single IP in e.Addresses, and that is why the logic is over complicated? Or, is there something missing in the understanding? Is there a way to make this a single slice.EndPoint with multiple addresses in the design?

Reply:
Hi @howardjohn and @kfaseela, please see below snapshot:
![image](https://github.com/istio/istio/assets/4101246/c2afa158-5848-44f9-9464-b7514724a2d4)
a dual stack pod contains 2 ip addresses and these 2 IP addresses are separated into 2 endpointslices based on type of IP family. Basically, each endpointslice contains 1 IP address with specified IP family. So there would be no 4 addresses in field `Addresses` of istioEndpoint.

Reply:
@zhlsunshine : please add a code comment clarifying the behaviour so that later on it won't confuse people. Also please add tests showing this works as expected.

Reply:
Hi @kfaseela beside above the modification file list:

1. pilot/pkg/serviceregistry/serviceentry/controller_test.go
2. pilot/pkg/serviceregistry/serviceentry/conversion_test.go
3. pilot/pkg/serviceregistry/serviceentry/store_test.go
4. pilot/pkg/model/push_context_test.go
5. pilot/pkg/model/validation_test.go

I also add more unit tests for multiple addresses change, please check the modification file list below:

```
modified:   pilot/pkg/networking/apigen/apigen_test.go

1.         modified:   pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go
2.         modified:   pilot/pkg/networking/core/v1alpha3/listener_builder_test.go
3.         modified:   pilot/pkg/networking/core/v1alpha3/listener_test.go
4.         modified:   pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer_test.go
5.         modified:   pilot/pkg/networking/grpcgen/grpcgen_test.go
6.         modified:   pilot/pkg/serviceregistry/kube/controller/controller_test.go
7.         modified:   pilot/pkg/serviceregistry/kube/controller/endpointslice_test.go
8.         modified:   pilot/pkg/xds/eds_test.go
9.         modified:   pilot/pkg/xds/endpoints/ep_filters_test.go
```

Reply:
Hello Steve, I would suggest to move this logic inside buildIstioEndPoint() so that it autmoatically gets covered by the tests. And then please add a new test with the feature flag enabled and we can verify the expected behavior